### PR TITLE
Made download_blogs requests async

### PR DESCRIPTION
**Pull request by cantudo**
Hello, I saw your message about issue #18 on Hacktoberfest's Discord. I have tried to address it, I have set tokio as a non optional dependency.

Using this feeds.txt:
```
https://antoniosbarotsis.github.io/index.xml
https://blog.rust-lang.org/feed.xml
https://blog.rust-lang.org/inside-rust/feed.xml
https://github.blog/feed
```
This is the output of hyperfine:
- With the synchronous requests:

image

- New asynchronous requests:

image

Please feel free to correct anything, this is my first time working with async and tokio in rust.

It would be helpful if you posted your feeds.txt, so i could test it more thoroughly. Thanks.

### CODE REVIEW

**Comment by AntoniosBarotsis** [comment]
Oh wow that was quick, here you go

```txt
https://antoniosbarotsis.github.io/index.xml

# Youtubers
https://www.youtube.com/feeds/videos.xml?channel_id=UCiSIL42pQRpc-8JNiYDFyzQ # Chris Biscardi
https://www.youtube.com/feeds/videos.xml?channel_id=UCUMwY9iS8oMyWDYIe6_RmoA # NoBoilerplate
https://www.youtube.com/feeds/videos.xml?channel_id=UC8ENHE5xdFSwx71u3fDH5Xw # ThePrimeagen
https://www.youtube.com/feeds/videos.xml?channel_id=UCsBjURrPoezykLs9EqgamOA # Fireship
https://www.youtube.com/feeds/videos.xml?channel_id=UC2Xd-TjJByJyK2w1zNwY0zQ # Beyond Fireship

# Rust stuff
https://blog.rust-lang.org/feed.xml
https://blog.rust-lang.org/inside-rust/feed.xml
https://this-week-in-rust.org/rss.xml
https://rust.libhunt.com/newsletter/feed
https://rustsec.org/feed.xml

# Console.dev 
https://console.dev/tools/rss.xml
https://console.dev/betas/rss.xml
https://console.dev/articles/rss.xml
https://console.dev/interviews/rss.xml
https://console.dev/profiles/rss.xml
https://blog.console.dev/rss/

https://blog.jetbrains.com/feed/
https://github.blog/feed/
https://stackoverflow.blog/feed/

# Random
https://dusted.codes/feed/rss
https://vladmihalcea.com/blog/feed/
https://arraying.de/index.xml
https://ossinsight.io/blog/rss.xml
https://www.smartercode.io/feed
https://fasterthanli.me/index.xml
https://raphlinus.github.io/feed.xml
https://www.huy.rocks/rss.xml
https://gideonwolfe.com/index.xml
https://blog.m-ou.se/index.xml
https://liam.rs/index.xml
https://qristin.wordpress.com/feed/
https://wakatime.com/blog/all.atom
https://developerlife.com/feed.xml

# Other
https://grafana.com/blog/index.xml
https://www.elastic.co/blog/feed
https://redis.com/blog/feed
https://www.stephanboyer.com/rss
https://medium.com/feed/@clockwork-labs
```

**Comment by AntoniosBarotsis** [comment]
It seems to be a bit slower than the master branch (for me at least)

```sh
# Yours
Benchmark 1: cargo run
  Time (mean ± σ):     20.506 s ±  3.317 s    [User: 3.331 s, System: 0.333 s]
  Range (min … max):   16.724 s … 22.921 s    3 runs

# Master
Benchmark 1: cargo run
  Time (mean ± σ):     14.180 s ±  0.784 s    [User: 4.654 s, System: 0.145 s]
  Range (min … max):   13.630 s … 15.078 s    3 runs
```

**Comment by cantudo** [comment]
Oh, thank you for your feedback, I am testing it also, but I get varying results, not sure why, maybe ratelimiting or something.

**Comment by AntoniosBarotsis** [comment]
I ran mine a few times and I seem to be getting roughly the same as the ones I posted (for the big input).

It seems that the pipeline is failing, this seems to be an issue with OpenSSL and can likely be solved by adding

```toml
[target.'cfg(unix)'.dependencies]
openssl = "0.10.42"
```

to `cargo.toml`. See [here](https://github.com/seanmonstar/reqwest#requirements).

**Comment by cantudo** [comment]
Without hyperfine I can see that there are some problems with the code, I am trying to fix it.

image

**Comment by AntoniosBarotsis** [comment]
This is definitely something introduced with the new HTTP library you added, I didn't get these in master.

**Comment by AntoniosBarotsis** [comment]
I didn't try this but it seems the issue is [here](https://github.com/cantudo/Rss2Email/blob/1bad6ee2060aa85c8e3c75d85f610e13929bbacd/src/lib.rs#L159-L167). The inputs include  say `application/xml` but the error is thrown because of `application/xml; charset UTF-8`. Maybe it would make more sense to do an includes check instead of checking for equality.

**Comment by cantudo** [comment]
Yest, it seems that reqwest shows the content type header in this format, `application/rss+xml; charset=UTF-8` when converting to a `&str`, maybe we could parse it so only the first part (before the semicolon) is passed into `is_supported_content` function. I am open to trying other libraries, i chose this one because it seemed well supported.

**Comment by AntoniosBarotsis** [comment]
I guess we could also try splitting, either works so up to you. The library itself is probably fine, it's just a matter of getting the string to cooperate 😄

**Comment by cantudo** [comment]
Now I am getting good results:

Benchmark 1: cargo run
  Time (mean ± σ):      5.279 s ±  2.544 s    [User: 4.914 s, System: 0.285 s]
  Range (min … max):    3.593 s …  8.205 s    3 runs

**Comment by AntoniosBarotsis** [comment]
I am getting all sorts of errors.

The requests seem to be too quick for my laptop to keep up

```
Reqwest(reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("grafana.com")), port: None, path: "/blog/index.xml", query: None, fragment: None }, source: hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 10060, kind: TimedOut, message: "A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond." })) })
```

I am also getting this?

```
Error in https://blog.jetbrains.com/feed/
    ["Syntax: 4:7 Unexpected closing tag: HEAD, expected META"]
```

I got that by adding `"text/html"` to the supported methods.


I am not sure if these issues are because of my internet connection, OS or laptop but considering you have been getting better results, it's probably not the code.

The extra dependency doesn't seem to be working, as I mentioned above;

> maybe it's just that OpenSSL is installed in the Github Runners and not in the Docker image

Maybe adding `openssl` [here](https://github.com/cantudo/Rss2Email/blob/32a83c7c711397cccd6ffa2f58b85e43ac090387/Dockerfile#L9) would fix it but I can just experiment with this later as it's not necessarily important for the scope of the PR.


Considering that this seems to have external issues that we can't control, would it make sense to make this a feature? Similar to how the `aws-lambda` feature works.

**Comment by cantudo** [comment]
Ahhh, I still think I can fix this. It is possible to limmit the number of concurrent connections, see the first answer [here](https://stackoverflow.com/questions/51044467/how-can-i-perform-parallel-asynchronous-http-get-requests-with-reqwest),

I went with
```let contents = rt.block_on(futures::future::join_all(...```
which makes all requests on the feeds.txt file concurrent, I used this because it was faster but I think we can do better and get it working on all machines. I will try again tomorrow and let you know

The jetbrains site has something weird, I think we could add a request header for the accepted content types like [this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept), maybe it will fix it, again I will check tomorrow.

Regarding openssl, in my opinion the best we can do is move to another async http client implementation, see [hyper](https://github.com/hyperium/hyper), using unlike reqwests can use rust-tls, I think this is better because of the reduced number of external dependencies.

Thanks for your help, I am sorry you could not see the results I was getting.

**Comment by AntoniosBarotsis** [comment]
Great to hear! I was also about to go as it is getting late here as well.

> I am sorry you could not see the results I was getting.

If it works, it works. I am hosting this on Linux anyway, Windows is weird sometimes. Even if we can't make this better on Windows, it will still be faster for 2 other major platforms and that's still useful.

Thanks a lot for your contributions!

**Comment by cantudo** [comment]
Hello again, I have fixed the jetbrains page issue, they were rejecting requests without a UserAgent header, so i just added a mock one. I also added `hyper-tls` as a feature for the reqwest crate, I am hoping this fixes de failing pipeline. I also limited the number of concurrent requests to 10, hopefully it is working on your machine this time. This is the last benchmark I ran with all the new changes and your feeds.txt:
```
# New code
Benchmark 1: cargo run
  Time (mean ± σ):      5.367 s ±  0.453 s    [User: 3.769 s, System: 0.034 s]
  Range (min … max):    5.079 s …  5.889 s    3 runs

# Master
Benchmark 1: cargo run
  Time (mean ± σ):     20.572 s ±  3.694 s    [User: 3.922 s, System: 0.026 s]
  Range (min … max):   17.919 s … 24.792 s    3 runs
```

**Comment by AntoniosBarotsis** [comment]
Don't worry about it, I just defined an absurd amount of rules. I'll review this later today, thanks for the commits!

**Comment by AntoniosBarotsis** [comment]
I am getting this now

```sh
Benchmark 1: cargo run
  Time (mean ± σ):      5.037 s ±  0.052 s    [User: 4.535 s, System: 0.293 s]
  Range (min … max):    4.977 s …  5.069 s    3 runs
```

so I guess you did manage to make this work on windows as well after all, good job!

I'll try this on AWS real quick and if that works fine (which it should since the pipeline is passing) I'll merge.

**Comment by AntoniosBarotsis** [comment]
Looking at the code, I think I actually managed to make this work yesterday when I was trying stuff out on my own before opening the issue.

I had tried this general pattern of `block_on(stream())` but I didn't specify `buffer_unordered(CONCURRENT_REQUESTS)` and I was getting times similar to what I was getting with your implementation earlier. I was just unaware that this was indeed faster and my laptop was just choking.

**Comment by AntoniosBarotsis** [comment]
I just have a random question about `CONCURRENT_REQUESTS`, is there a reason you picked static over const? I am relatively new to Rust so not too sure what the difference is.

**Comment by AntoniosBarotsis** [comment]
Ok, idk how but this is giving me segfaults on AWS Lambda 🗿🗿🗿

Any ideas?

**Comment by cantudo** [comment]
I dont have my laptop rn but I think it might be the tokio initialization, if I am not mistaken AWS runs with the tokio::main, which already creates a runtime, so creating the rt variable fails in AWS, I think we should either make everything async in local mode o add a config # such as the one you have on main.rs.

**Comment by AntoniosBarotsis** [comment]
Wym `async in local mode`?

Do you mean that we should make everything up to the main function async?

**Comment by cantudo** [comment]
yeah, i think its a matter of just making download_blogs and main async, but i have not tested it

**Comment by cantudo** [comment]
Hello again, I have tried to implement a quick fix, could you try it on lambda? it now checks to see if there already exist a tokio runtime, to not create it all over again.

> I've seen people say stick to const whenever possible if no mutability is involved so I was just wondering.
You are right, I am also relatively new to rust, this was a genuine mistake I have set it to const now.

> Would your implementation work if that was the case since you call it from inside an async block anyway?
I do not think so, since async/await is not multithreading (everything runs on one thread using a scheduler) calling a synch request would halt everything until it finished, making everythin "synch" even if called from asynch, because the scheduler does not now that it should pass on to other stuff while waiting for IO on that call (since it is not asynch).

**Comment by AntoniosBarotsis** [comment]
I am still getting a seg fault :/

**Comment by cantudo** [comment]
Oh, okay, do you know where it fails?

**Comment by AntoniosBarotsis** [comment]
Unfortunately the error is just

```txt
Error: Runtime exited with error: signal: segmentation fault
```

not exactly insightful

**Comment by cantudo** [comment]
Damn haha, I am looking [here](https://github.com/awslabs/aws-lambda-rust-runtime/issues/270)

**Comment by AntoniosBarotsis** [comment]
bruh I somehow looked up "segfault" and "segmentation" earlier in that repo and didn't see this issue true

**Comment by cantudo** [comment]
Haha github searching is confusing for me, I find that google works better most times, currently testing docker build with the cargo.toml fix stated in that issue.

**Comment by AntoniosBarotsis** [comment]
I'll setting it to and gonna see if that works

```toml
reqwest = { version = "0.11", default-features = false, features = ["json", "rustls"] }
```

**Comment by cantudo** [comment]
Did it work? For me the openssl dependency is still there even after modifying cargo.toml

**Comment by AntoniosBarotsis** [comment]
Ok so I got 2 different errors;

```
Reqwest(reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("medium.com")), port: None, path: "/feed/@clockwork-labs", query: None, fragment: None }, source: hyper::Error(Connect, "invalid URL, scheme is not http") })
```

and the second time I ran it

```
{
  "errorType": "&alloc::boxed::Box<dyn core::any::Any+core::marker::Send>",
  "errorMessage": "Lambda panicked"
}


thread 'main' panicked at 'Builder::init should not be called after logger initialized: SetLoggerError(())', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/env_logger-0.9.0/src/lib.rs:820:14
```

**Comment by cantudo** [comment]
I dont know about the first error, like you say it seems that it would happen in local, but I have not seen it on my machine. Hyper should know that the requests are https, it may be something with the dependencies, can you test the new changes?.
Regarding the second error, I have modified the env_logger init to try_init, I have not touched, I have also removed the openssl crate from cargo.toml (which we still had from yesterday).

**Comment by AntoniosBarotsis** [comment]
Ok this seems to be working, I'll check some things later again because I grew paranoid and I'll likely merge tonight, huge thanks!

**Comment by cantudo** [comment]
Test it well please! I would feel terrible if this caused issues, only wanted to help.

**Comment by AntoniosBarotsis** [comment]
I ran this 3-4 times already but it seems to be fine, the emails are sent, and no errors.

Performance went from 18 to 11 seconds which is great!

**Comment by cantudo** [comment]
Glad to hear that! But I am still having issues:

```
[2022-10-09T20:17:33Z WARN  rss2email] Error in https://developerlife.com/feed.xml
    Reqwest(reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("developerlife.com")), port: None, path: "/feed.xml", query: None, fragment: None }, source: hyper::Error(Connect, Custom { kind: Other, error: Custom { kind: UnexpectedEof, error: "tls handshake eof" } }) })
```

**Comment by AntoniosBarotsis** [comment]
I don't seem to be getting this locally 🤔

**Comment by cantudo** [comment]
The hell, maybe it is me, I am going to clean cargo cache and test this again.

**Comment by AntoniosBarotsis** [comment]
Also what I don't understand, did the logger make this crash??? Or was it the other stuff you changed in cargo.toml

**Comment by AntoniosBarotsis** [comment]
Turned into a draft so I don't click big green merge button accidentally

**Comment by cantudo** [comment]
> Also what I don't understand, did the logger make this crash??? Or was it the other stuff you changed in cargo.toml

I think it was the logger but we could try to revert that line and see if it crashes

> Turned into a draft so I don't click big green merge button accidentally

Good idea haha

**Comment by cantudo** [comment]
I think the problem is with my computer, I am getting this error now on your master :face_exhaling: 

```
2022-10-09T20:38:04Z WARN  rss2email] Error in https://antoniosbarotsis.github.io/index.xml
    Ureq(Transport(Transport { kind: ConnectionFailed, message: Some("tls connection init failed"), url: Some(Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("antoniosbarotsis.github.io")), port: None, path: "/index.xml", query: None, fragment: None }), source: Some(Kind(UnexpectedEof)) }))
```

**Comment by AntoniosBarotsis** [comment]
Why did this become so cursed. Both branches work fine for me now.

??????????????????

**Comment by cantudo** [comment]
I do not now hahaha I am removing all cargo packages and installing again, just to be sure.

**Comment by cantudo** [comment]
I have tested this on another machine and it also works for me, I believe I messed something up in my rust installation while testing. If you still want to merge I would recommend further testing, maybe a friend or something because I cannot confidently say this will not fail.

**Comment by AntoniosBarotsis** [comment]
Appreciate the concern ♥

I will force my friends to run this tomorrow after running it myself a few times.

**Comment by AntoniosBarotsis** [comment]
I'll try to get at least one person to run the code on each platform, are you using Mac or Linux?

**Comment by AntoniosBarotsis** [comment]
The pipeline passing is definitely reassuring here, made it build, test and run the code in the same `feeds.txt` that we are both using in this issue on all major platforms.

I think I will wait until #20 is resolved at least partially (just the extra Arm image part) and if that seems to also be working without issues then I'd say it is safe to assume this is working fine and I'll finally merge.

**Comment by AntoniosBarotsis** [comment]
The GitHub Runners have pretty fast internet connections so the speedup is smaller (from 10 to 7 seconds) but that is expected and looks fine to me. It also seems to work fine even if ran multiple times (the benchmark runs the code 5 times) which is good to know.

Are you still getting those errors locally?

**Comment by cantudo** [comment]
>  It also seems to work fine even if ran multiple times (the benchmark runs the code 5 times) which is good to know.

That is great to hear!

> Are you still getting those errors locally?

Unfortunately, yes, at this point I think my internet connection is just bad, I tested with my uni connection this morning and everything ran well, I also tested it in a uni server and it ran well. Running the current master version (still with ureq) it gives me the same errors at home, but if I activate my uni vpn it works without a problem. This is two runs of the current master, one without and one with VPN.

```
# First run (current master no async and no vpn)
2022-10-10T20:14:08Z INFO  rss2email] Days set to 7
[2022-10-10T20:15:11Z WARN  rss2email] Error in https://blog.rust-lang.org/feed.xml
    Ureq(Transport(Transport { kind: ConnectionFailed, message: Some("tls connection init failed"), url: Some(Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("blog.rust-lang.org")), port: None, path: "/feed.xml", query: None, fragment: None }), source: Some(Kind(UnexpectedEof)) }))
^C⏎
# Second run (current master no async and using my university's vpn)
cantu@cantu-l ~/U/T/Rss2Email (master) [SIGINT]> cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/rss2email`
[2022-10-10T20:15:50Z INFO  rss2email] Days set to 7
[2022-10-10T20:16:31Z INFO  rss2email] Elapsed time for download_blogs was 40561ms
[2022-10-10T20:16:31Z INFO  rss2email] Downloaded 17 blogs with 97 posts total.
[2022-10-10T20:16:31Z INFO  rss2email] <h1>Rss2Email - 2022-10-10UTC</h1><h2>Tony's Blog</h2><ul
```

Sometimes it runs without errors and sometimes I get those, not always on the same domains. I have also tested this at home with a Raspberry Pi 3 and the same thing happens. At this point I just do not know what happens, but whatever it is I think is beyond my control.

**Comment by AntoniosBarotsis** [comment]
Yes seems like it's your wifi then.

I didn't find much regarding the error with one Google search but I'll ask around in case anyone has an idea about this out of curiosity.

This again means that the PR is very likely fine but I'll wait as I mentioned above regardless. Worst case scenario, if it is indeed buggy then I'll have more time to encounter the issue.

**Comment by cantudo** [comment]
> I didn't find much regarding the error with one Google search but I'll ask around in case anyone has an idea about this out of curiosity.

Great, let me know if you fins something out!

> This again means that the PR is very likely fine but I'll wait as I mentioned above regardless. Worst case scenario, if it is indeed buggy then I'll have more time to encounter the issue.

Sure, sounds good also I am sorry this took so long, in my mind it was a simple change. I do not study / work in the software engineering field so this was definitely a learning experience for me, I have loved seeing the CI workflow you have implemented, it works really well.

**Comment by AntoniosBarotsis** [comment]
> I am sorry this took so long

I really do not mind especially seeing as how I would likely not do this myself, at least not now.

> this was definitely a learning experience for me

That is great to hear :)
Being new to Rust I have also learned a lot of stuff through this project

File: src/lib.rs [hunk 1]
```diff
@@ -1,51 +1,83 @@
...
 use std::fmt::Write as _;
+use tokio::runtime::Handle;
 
 pub use blog::{Blog, Post};
 mod blog;
 mod xml;
 
 use crate::xml::parse_web_feed;
 
+const CONCURRENT_REQUESTS: usize = 10;
```
Comment (AntoniosBarotsis) [inline src/lib.rs:19]: What was the thought process for arriving at this value? Were other values considered?

After trying `5` and `15`, `10` seems to be working better on my machine, just curious.

**Review by AntoniosBarotsis — CHANGES_REQUESTED** [review]

**Comment by AntoniosBarotsis** [comment]
I have tested the code manually as well and it seems fine, performance went from 20 to 12 seconds!

File: src/lib.rs (hunk 1 above)
Comment (cantudo) [reply src/lib.rs:19]: Yes, I did try various sizes, 10 seemed a good in between, lower were getting me slower times and higher would risk having the problem you had on Windows when all were concurrent.

File: src/lib.rs [hunk 2]
```diff
@@ -153,6 +204,49 @@ pub fn get_page(url: &str) -> Result<String, DownloadError> {
...
+        Err(_) => Err(DownloadError::Custom(
+          "Content Type parsing error".to_string(),
+        )),
+      }
+    }
+    None => Err(DownloadError::Custom(
+      "No content type header found on request.".to_string(),
+    )),
+  }
+}
```
[the remark this answers is not in the document]
Comment (cantudo) [reply src/lib.rs]: It looks good! Much more readable.

File: src/lib.rs (hunk 1 above)
Comment (AntoniosBarotsis) [reply src/lib.rs:19]: Makes sense

**Review by AntoniosBarotsis — APPROVED** [review]

**Comment by AntoniosBarotsis** [comment]
Thanks a lot!

**Pull request merged**